hms_tz/nhif/api/patient_encounter.py

```diff
@@ -86,62 +122,92 @@ def on_submit_validation(doc, method):
+        "Healthcare Service Insurance Coverage",
+        fields={
+            "healthcare_service_template",
+            "maximum_number_of_claims",
+            "approval_mandatory_for_claim",
+        },
+        filters={
+            "is_active": 1,
+            "healthcare_insurance_coverage_plan": healthcare_insurance_coverage_plan,
+            "start_date": ("<=", nowdate()),
+            "end_date": (">=", nowdate()),
+            "healthcare_service_template": ("in", healthcare_service_templates),
+        },
+        order_by="modified desc",
+    )
+
+    hsic_map = {hsic.healthcare_service_template: hsic for hsic in hsic_list}
+    for template in healthcare_service_templates:
+        if template not in hsic_map:
+            msgThrow(
+                _("{0} not covered in Healthcare Insurance Coverage Plan").format(
+                    template
+                ),
+                method,
+            )
+            continue
+
+        coverage_info = hsic_map[template]
+        if coverage_info.maximum_number_of_claims == 0:
+            continue
+
+        times = 12 / coverage_info.maximum_number_of_claims
+        count = 1
+        days = int(times * 30)
+        if times < 0.1:
+            count = 1 / times
+            days = 30
+
+        if not coverage_info.number_of_claims:
+            coverage_info.number_of_claims = frappe.db.count(
+                "Healthcare Insurance Claim",
+                {
+                    "service_template": template,
                     "insurance_subscription": insurance_subscription,
-                    "claim_posting_date": ["between", start, end],
-                })
-                if len(claims_count) > count:
-                    frappe.throw(_("Maximum Number of Claims for {0} per year is exceeded within the last {1} days").format(
-                        row.get(value), days))
-                row.is_restricted = next(i for i in hsic_list if i["healthcare_service_template"] == row.get(
-                    value)).get("approval_mandatory_for_claim")
-                if row.is_restricted:
-                    frappe.msgprint(_(row.doctype + " row " + str(row.get(value)) +
-                                      " requires additional authorization"), alert=True)
+                    "claim_posting_date": ["between", add_to_date(today, -days), today],
+                },
+            )
+
+        if coverage_info.number_of_claims > count:
+            frappe.throw(
+                _(
+                    "Maximum Number of Claims for {0} per year is exceeded within the"
+                    " last {1} days"
+                ).format(template, days)
+            )
+
+        for row in healthcare_service_templates[template]:
+            row.is_restricted = coverage_info.approval_mandatory_for_claim
+            if row.is_restricted:
+                frappe.msgprint(
+                    _("{0} with template {1} requires additional authorization").format(
```

Comment:
I have slightly changed this error message. We can also used `idx` for row number here.

Reply:
Acceptable